### ISSUE TITLE
[dev] Graphical Corruption in TurboVNC Viewer on Windows

### ISSUE THREAD

**Issue by Parakleta** [TurboVNC/turbovnc#254]
See the following image.

![TurboVNC corruption](https://user-images.githubusercontent.com/11205202/101422962-0f2f7c00-394c-11eb-89d3-73d97e1b22b9.png)

Interestingly if I set the scaling to 105% instead of 100% the problem goes away.  Also I can take a screen shot using TurboVNC itself (using the "Save remote desktop image") and the screenshot is perfect.

This is using build v2.2.80 (20201119).  Note that I wasn't able to install TurboVNC directly because windows flags the installer as a virus, instead I had to unpack it on a unix machine and copy the viewer folder to windows, in case that makes any difference.

**Comment by dcommander** [comment]
Does either issue (the virus detection) or the image corruption occur in the 2.2.5 stable release?

**Comment by dcommander** [comment]
Also please specify which AV software is flagging our installer.  I need to be able to reproduce that myself.

**Comment by Parakleta** [comment]
2.2.5 stable is fine.

Windows Smartscreen blocks any of the CI builds that I have tested.  Actually, now that I try again it will let me override and keep it.  Previously it was flagging it as having 'Trojan:Win32/Spursint.F!cl'.  I'll try again using the installer and see if the glitch still exists.

**Comment by Parakleta** [comment]
Install succeeds now, but the graphical glitches remain.

Using v2.2.80 (20201208) now.

**Comment by Parakleta** [comment]
I hadn't realised 2.2.5 had both the Java and non-Java versions.  The 2.2.5 non-Java is fine, the 2.2.5 Java show the same corruptions.

**Comment by Parakleta** [comment]
I have done some further testing and discovered that it is due to the "Scale and layout" settings in Windows.  I am using a laptop with a high DPI screen so windows uses 125% UI scaling, but when I connect a regular external monitor with 100% UI scaling the VNC graphics are all fine.

**Comment by dcommander** [comment]
OK, I'll try to reproduce and come up with a workaround.  Thanks for testing!

**Comment by dcommander** [comment]
I updated the version of Inno Setup used in both the CI and official builds to 6.1.2 (previously 5.5.x), so maybe that will help with the MS Defender false positive.  I was never able to reproduce that issue, though.

As far as the scaling issue, when display scaling is enabled in Windows, applications (including the native Windows TurboVNC Viewer) that use Win32 API functions like `GetClientRect()`, `GetWindowRect()`, or `EnumDisplayMonitors()` are given fake screen and window coordinates/dimensions that make it look like the display is smaller than it actually is (20% smaller along both screen dimensions, in the case of 125% scaling.)  When automatic desktop resizing is enabled in the TurboVNC Viewer, which it is by default, that makes the remote desktop 20% smaller along both screen dimensions as well, since the TurboVNC Viewer resizes the remote desktop based on the fake window dimensions.

In Java/Swing, apparently it was necessary to add special support for DPI awareness.  Java 8 doesn't have it, but Java 9 and later does.  Unfortunately, however, the scaling algorithm it uses has some issues:
https://bugs.openjdk.java.net/browse/JDK-8187585
Until those issues are resolved, the only workarounds I know of are:
1. Use the native Windows TurboVNC Viewer.
2. Disable DPI awareness in Java by setting the `JAVA_TOOL_OPTIONS` environment variable to `-Dsun.java2d.uiScale.enabled=false`.  That will restore the behavior of Java 8, but of course that means that, if you're using a high DPI monitor, your remote desktop will appear smaller than you might expect.

**Comment by Parakleta** [comment]
Thanks.  It seems the behaviour varies quite significantly with different Java versions.  At one point my computer decided to change which version of Java was launching the `.jar` to JDK8 and it just started working on my laptop display at 125% scaling, but then whenever I tried to move it onto the external monitor it scaled the video in reverse (i.e. a 1600x900 external screen was being perceived as 2000x1125).

Launching it with JDK11 and using the switch you provided gives acceptable behaviour.  Thanks.

**Comment by dcommander** [comment]
I can easily change the default value of that flag within the TurboVNC Viewer application.  I'm just not sure yet whether that's the right thing to do.  I need to experiment with an actual high DPI monitor and see what makes the most sense.

**Comment by dcommander** [comment]
This should be worked around in the latest 3.0 post-beta pre-release build.
https://turbovnc.org/DeveloperInfo/PreReleases

**Comment by Parakleta** [comment]
That seems to do it.  I appreciate the lengths you've gone to to try and fix this.  The more I deal with (and see other people dealing with) issues in Java the more frustrated I become with how clumsy it all is.

It's hilarious (in a sad way) to me that Java which was supposed to run anywhere is now devolving into a situation where people are having to pack their own JRE with their applications just to get it to work reliably.  I've had a disastrous time trying to get Java programs to run on the new Apple M1 devices.

As I was playing around with this I was thinking about the zoom setting that you already have in the viewer.  Is there any way to disable the built in scaling and use your zoom setting to compensate?  The scaling only seems to go wrong for the image rendering space, since the UI widgets all seem to work fine.  Interestingly this new behaviour actually makes the widgets look worse:

Without new scaling behaviour:
![TurboVNC_noscale](https://user-images.githubusercontent.com/11205202/161455240-46babc7a-9394-4a41-9a13-fd1b7de4f83e.png)

With new scaling behaviour:
![TurboVNC_scale](https://user-images.githubusercontent.com/11205202/161455169-a83a36a8-19bf-418f-8215-2940d635b61b.png)

Someone on stackoverflow claims that the UI Scaling can be disabled on a single Graphics2D object here https://stackoverflow.com/questions/43057457/jdk-9-high-dpi-disable-for-specific-panel, maybe this could work?

**Comment by dcommander** [comment]
> That seems to do it. I appreciate the lengths you've gone to to try and fix this. The more I deal with (and see other people dealing with) issues in Java the more frustrated I become with how clumsy it all is.
> 
> It's hilarious (in a sad way) to me that Java which was supposed to run anywhere is now devolving into a situation where people are having to pack their own JRE with their applications just to get it to work reliably. I've had a disastrous time trying to get Java programs to run on the new Apple M1 devices.

No one ever accused Sun Microsystems of focus or consistency with regard to their product strategies.  IMHO, with respect to both Solaris and Java, they could never seem to decide whether they wanted those technologies to be open standards or cash cows.  By not choosing a path and sticking with it, they effectively closed both paths.  (In the immortal words of Neil Peart, "If you choose not to decide, you still have made a choice.")  Java is a great language, but Sun tried to rule over it with an iron fist.  They didn't make it open enough to encourage buy-in from other large tech companies, but they also didn't invest enough of their own resources to keep Java abreast of the latest advancements in graphics and security.  Because Sun refused to update the Java plugin, which used the outdated and insecure NPAPI standard, a lot of IT departments started banning Java entirely, which undercut Java Web Start. (JWS only makes sense if all potential users have a JRE installed.)  Once JREs were no longer ubiquitous, then there was little choice but for Java application developers to embed their own JREs.  I personally believe that, had Sun created an OpenJDK-like project in 2005, every operating system installation in 2022 would have a JRE by default.  Instead, they doubled down on their proprietary technologies rather than reinventing themselves around open technologies like IBM did.  They focused too much on selling milk from the cash cow and not enough on the quality of the milk.  So it goes.  At this point, OpenJDK is essentially just providing a FOSS path forward for those who have invested too much in Java to abandon it, but there are major issues with it that seem to get no attention at all from the developers.  It's notable that, despite being an "open" project, there is still no way for the average user to file a bug report against OpenJDK.

With respect to TurboVNC, for the last 12 years I have wanted to develop a cross-platform native viewer using a modern GUI framework such as Qt or GTK, but I have never been able to secure enough funding to do that.  The funding I was able to secure was toward improving the Java TurboVNC Viewer, because at the time (early 2010s), applets and JWS were still a thing, and several of my paying customers had embraced those technologies for implementing a zero-install remote desktop solution.  As a result, the Java TurboVNC Viewer had by far the most momentum and was thus in the best position to become the "One TurboVNC Viewer to Rule Them All" in TurboVNC 3.0.  However, if someone paid me to overhaul the viewer using C++, I would do it.  As the Java viewer evolves, I am forced to use JNI more and more to work around deficiencies in Swing, so the primary purpose of Java from TurboVNC's point of view is to handle network communications and draw a GUI.  The Java viewer is much easier to maintain than the old native Windows TurboVNC Viewer, which was written using raw Win32 GUI code, but that's not saying much.  There is no technological reason for us to use Java anymore.  The only reason we continue to use it is that it would take hundreds of hours to switch to something else.

> As I was playing around with this I was thinking about the zoom setting that you already have in the viewer. Is there any way to disable the built in scaling and use your zoom setting to compensate? The scaling only seems to go wrong for the image rendering space, since the UI widgets all seem to work fine. Interestingly this new behaviour actually makes the widgets look worse:
> 
> Without new scaling behaviour: ![TurboVNC_noscale](https://user-images.githubusercontent.com/11205202/161455240-46babc7a-9394-4a41-9a13-fd1b7de4f83e.png)
> 
> With new scaling behaviour: ![TurboVNC_scale](https://user-images.githubusercontent.com/11205202/161455169-a83a36a8-19bf-418f-8215-2940d635b61b.png)
> 
> Someone on stackoverflow claims that the UI Scaling can be disabled on a single Graphics2D object here https://stackoverflow.com/questions/43057457/jdk-9-high-dpi-disable-for-specific-panel, maybe this could work?

\<facepalm\>  I was apparently making this a lot more difficult than it needed to be.  I didn't realize that I can actually override the high-DPI scaling algorithm.  Can you try reverting 833967be2aa0ba45e1cef26efb7bd7db25f88d20 and then applying the following?

```
--- a/java/com/turbovnc/vncviewer/DesktopWindow.java
+++ b/java/com/turbovnc/vncviewer/DesktopWindow.java
@@ -1,7 +1,7 @@
 /* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
  * Copyright (C) 2006 Constantin Kaplinsky.  All Rights Reserved.
  * Copyright (C) 2009 Paul Donohue.  All Rights Reserved.
- * Copyright (C) 2010, 2012-2013, 2015-2018, 2020-2021 D. R. Commander.
+ * Copyright (C) 2010, 2012-2013, 2015-2018, 2020-2022 D. R. Commander.
  *                                                     All Rights Reserved.
  * Copyright (C) 2011-2013 Brian P. Hinz
  *
@@ -411,6 +411,10 @@ class DesktopWindow extends JPanel implements Runnable, MouseListener,
       super.paintComponent(g);
     if (cc.viewport != null && (cc.viewport.dx > 0 || cc.viewport.dy > 0))
       g2.translate(cc.viewport.dx, cc.viewport.dy);
+    double scalingFactor = g2.getTransform().getScaleX();
+    if (scalingFactor != 1.0)
+      g2.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
+                          RenderingHints.VALUE_INTERPOLATION_BILINEAR);
     if (cc.cp.width != scaledWidth || cc.cp.height != scaledHeight) {
       g2.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
                           RenderingHints.VALUE_INTERPOLATION_BILINEAR);
```

**Comment by dcommander** [comment]
Never mind.  I went ahead and pushed the fix temporarily so others can test it.  You can test it in the latest 3.0 post-beta pre-release build for Windows: https://turbovnc.org/DeveloperInfo/PreReleases.  Hopefully what you will observe is that the toolbar and remote desktop area look the same as with the previous fix, but the text-based widgets look the same as with 3.0 beta1.

**Comment by Parakleta** [comment]
It is a thing of beauty... well done.  It even seems to look less blurry than using windows built in scaling.

> There is no technological reason for us to use Java anymore.

Frustratingly it is also still widely taught.  I teach programming in a college and I'm currently forced to teach it due to the state legislated curriculum.  It was introduced when applets were seen as an easy way to get a program up and running anywhere, and sadly we're still required to teach applets by the curriculum although now it is done entirely in theory since there's no practical way to run them anymore.  I think they also like to keep Java because then the IT guys can keep the computers locked down so that students cannot run arbitrary binaries (completely overlooking the fact that you can call into DLLs from interpreted/VM languages).  The course is due for redevelopment at the end of this year and I'm fighting hard to drop Java.  I think if I fail in that fight I might give up teaching and go back to programming...

**Comment by dcommander** [comment]
Out of curiosity, what would be your preferred language to teach beginners?  I get asked that question a lot.

**Comment by Parakleta** [comment]
There's not a simple answer to language choice because it depends on the audience's age and maturity and what they are hoping to get out of it.

If they are just hoping to kludge some things together and don't need to actually learn to program in the future Python would be fine, but I'm of the opinion that learning Python first is harmful to people who wish to develop actual programming skills in the future because it leads to a laziness of mind that is hard to overcome later.  Kind of like the whole "goto considered harmful" thing with Dijkstra.

Personally if they're really beginning and not yet at the point of wanting some real CS theory I would probably point them to Lua.  It's syntax and standard library is very small which means that they can master the language easily and start thinking about how to actually solve problems.  It also has a very simple meta-programming model which allows you to start adding power without much complexity.  Additionally the Love2D engine is fun to play with.

For a student wanting CS theory I'd lean towards a Lisp/Scheme/Racket type language so that they can really understand program design, with strong access to functional programming, lambda's, and meta-programming.  The http://htdp.org online course is good.  Unfortunately Racket has very clumsy structure syntax.  There's also the famous SICP book but its content is a bit mathematical/technical for modern students.

I really wish I could have a good type system in a first language as well, to force a rigour in program design, but in reality the shackles of an incomplete type system often prevent a lot of meta-programming which I think is more valuable to experience.  I'm tempted by languages like Haskell for the amazing type system and functional programming but I'm not entirely sure how not to scare off beginning students.  The http://www.learnyouahaskell.com book is a decent resource.

Typescript from Microsoft is also quite intriguing to me at the moment, since you get the flexibility of JavaScript with the safety of a type system, but you aren't strictly bound by it so you can break out when you need to do something clever.  I haven't really drilled in yet to figure out what hidden traps there might be however.  Typescript is available in their online coding environment http://makecode.com

In summary: Lua for most people just wanting to learn something, and then probably one of the functional languages for people wanting to understand CS.
